src/lib.rs

```diff
@@ -6,52 +7,194 @@ extern crate rand;
 
 use rand::{ChaChaRng, Rng, SeedableRng};
 use std::collections::HashMap;
+use std::rc::Rc;
+
+#[derive(Debug, Clone)]
+enum BitGenerator {
+    Random(ChaChaRng),
+    Recorded(Rc<Vec<u64>>),
+}
 
 #[derive(Debug, Clone)]
 struct DataSource {
+    bitgenerator: BitGenerator,
+    record: Vec<u64>,
+}
+
+#[derive(Debug, Clone)]
+enum Status {
+    Overflow,
+    Invalid,
+    Valid,
+    Interesting,
+}
+
+#[derive(Debug, Clone)]
+struct Engine {
     random: ChaChaRng,
+    max_examples: u64,
+    valid_examples: u64,
+    invalid_examples: u64,
+    interesting_examples: u64,
+    best_example: Option<Rc<Vec<u64>>>,
+}
+
+impl DataSource {
+    fn bits(&mut self, n_bits: u64) -> Option<u64> {
+        let mut result = match self.bitgenerator {
+            BitGenerator::Random(ref mut random) => random.next_u64(),
+            BitGenerator::Recorded(ref mut v) => if self.record.len() >= v.len() {
+                return None;
+            } else {
+                v[self.record.len()]
+            },
+        };
+
+        if n_bits < 64 {
+            let mask = (1 << n_bits) - 1;
+            result &= mask;
+        };
+
+        self.record.push(result);
+
+        return Some(result);
+    }
+
+    fn new(generator: BitGenerator) -> DataSource {
+        return DataSource {
+            bitgenerator: generator,
+            record: Vec::new(),
+        };
+    }
+
+    fn from_random(random: ChaChaRng) -> DataSource {
+        return DataSource::new(BitGenerator::Random(random));
+    }
+
+    fn from_vec(record: Rc<Vec<u64>>) -> DataSource {
+        return DataSource::new(BitGenerator::Recorded(record));
+    }
+}
+
+impl Engine {
+    fn new(max_examples: u64, seed: &[u32]) -> Engine {
+        return Engine {
+            random: ChaChaRng::from_seed(seed),
+            max_examples: max_examples,
+            valid_examples: 0,
+            invalid_examples: 0,
+            interesting_examples: 0,
+            best_example: None,
+        };
+    }
+
+    fn should_continue(&self) -> bool {
+        return (self.valid_examples < self.max_examples)
+            && (self.valid_examples + self.invalid_examples < self.max_examples * 10)
+            && (self.interesting_examples == 0);
+    }
+
+    fn mark_finished(&mut self, source: DataSource, status: Status) -> () {
+        match status {
+            Status::Overflow => (),
+            Status::Valid => self.valid_examples += 1,
+            Status::Invalid => self.invalid_examples += 1,
+            Status::Interesting => {
+                self.interesting_examples += 1;
+                self.best_example = Some(Rc::new(source.record));
```

Comment:
While the Rc here avoids copying the Vec (instead copying the reference), it also complicates the program a little, given that `Vec<u64>` is `Clone`, do you think it'd be preferential to clone the vec and not the Rc (this takes the performance hit of copying the data, which may be unacceptable, but I figured I'd point out the complexity tradeoff)

Reply:
Hmm. I was thinking we'd need to copy this more often than we actually do. Maybe this is premature optimisation. I'll take it out.